indexes for mysql schema dump

jwh315 opened the pull request:
I ran across your schema dump/load feature for phinx a few weeks ago.  I really love your implementation but really needed indexes.  I was able to get the indexes put into the mysql adapter pretty easily.  

I have a pretty decent size dev database (over 700 tables) at my company, that has an array of ridiculous design inconsistencies from years of development.  The indexing seems to be recreating tables 100% from their original state, so I am fairly confident that I have covered most edge cases related to indexing and phinx.

Not sure if you would be interested in merging this branch into yours, or if you would need more from my end before doing so.

Either way, thanks for the effort putting this schema dumper together, it has saved me several hours worth of work.

Commit:
add index capability to the schema:dump feature

Commit:
-set index limits where applicable
-fix some primary key issues which were cropping up on tables where a column name was id but was not the pk

basilfx commented:
Sounds like a great addition! Let me take a look at it.

basilfx commented on src/Phinx/Db/Adapter/AdapterInterface.php:
```diff
@@ -47,6 +47,7 @@
     const PHINX_TYPE_CHAR           = 'char';
     const PHINX_TYPE_TEXT           = 'text';
     const PHINX_TYPE_INTEGER        = 'integer';
+    const PHINX_TYPE_DOUBLE         = 'double';
```
I think this would be better in a separate PR.

basilfx commented on src/Phinx/Db/Adapter/MysqlAdapter.php:
```diff
@@ -857,6 +864,9 @@ public function getSqlType($type, $limit = null)
             case static::PHINX_TYPE_FLOAT:
                 return array('name' => 'float');
                 break;
+            case static::PHINX_TYPE_DOUBLE:
```
I think this would be better in a separate PR.

basilfx commented on src/Phinx/Db/Adapter/MysqlAdapter.php:
```diff
@@ -533,17 +534,22 @@ public function dropColumn($tableName, $columnName)
      * @return array
      */
-    protected function getIndexes($tableName)
+    public function getIndexes($tableName, $lowercase = true)
```
What's the added benefit of lowercase?

jwh315 replied:
this method was originally lowercasing all index column names (as it is currently).  I initially thought I needed to have case sensitive column names for mysql on some operating systems.  That turned out to be incorrect thinking on my part.  I have removed the extra param

basilfx commented:
Can you also rebase (interactively) to fixup 2d8ebb3 in 8904d29?

Commit:
-remove double feature type from this pull request
-remove lowercase param from index function, indexes are not case sensitive in mysql
-account for tables that do not have primary keys when dumping a schema

jwh315 commented:
I removed the DOUBLE column type and fixed the commit history to remove the superfluous commit.  I also noticed another issue that needed to be addressed for tables that do not have primary keys, that has been fixed in the most recent commit.

basilfx commented:
Looks good to me, thanks!

basilfx commented:
I made a [few improvements](https://github.com/basilfx/phinx/commit/1bb96f37fa1a9164d7289c4e2969f3f1221610e3) to the code:
- There was still a reference to double in the PDOAdapter
- Undefined index check
- Optimized template

Other than that, it works great :-)

futoase commented on src/Phinx/Migration/Schema/Schema.template.php:
```diff
@@ -15,6 +15,12 @@ public function change()
         <?php endforeach; ?>
+
+        <?php foreach($table->getIndexes() as $name => $index): ?>
+            <? if ($name !== 'PRIMARY'): ?>
```
``` php
<?php if ($name !== 'PRIMARY'): ?>
```

basilfx replied:
Fixed, thanks!

basilfx commented:
@jwh315 I have fixed some more:
- 84cc9b7d48ba5e4a377815ccb80befebbbc00ee8 reverts another change that should not be part of this fork.
- 926918f64559eee8cf2be42a0c99bcf8006541e0 column names should not be converted to lower case, just interpret them as returned by MySQL's `SHOW INDEXES`.
- 322eedf9e9affa0551a9d543456065e2f2daab4d fixes a problem with the code generation: in `buildIndexString` the `$name` parameter caused issues. In SQLite, the name of an index cannot match a column name. I remove this parameter, but added it back as an option of the `Index` class.
